File: backend/app/models/cash_reconciliation.py
import enum
import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import Date, Enum as SQLEnum, ForeignKey, Numeric, Text, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.types import DateTime

from app.models.base import TenantScopedBase


class ReconciliationStatus(str, enum.Enum):
    open = "open"
    closed = "closed"


class CashReconciliation(TenantScopedBase):
    __tablename__ = "cash_reconciliations"
    __table_args__ = (
        UniqueConstraint("tenant_id", "business_date", name="uq_cash_reconciliation_date"),
    )

    business_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    opening_balance: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0"))
    # cash_in, petty_cash_net, expected_balance are computed on read — not stored
    counted_balance: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    deposit_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, default=Decimal("0"))
    closing_balance: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    variance: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    variance_note: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[ReconciliationStatus] = mapped_column(
        SQLEnum(ReconciliationStatus, name="reconciliation_status"),
        nullable=False,
        default=ReconciliationStatus.open,
    )
    closed_by_user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    closed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class PettyCashEntry(TenantScopedBase):
    __tablename__ = "petty_cash_entries"

    reconciliation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("cash_reconciliations.id"), nullable=False, index=True
    )
    amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    created_by_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )

File: frontend/src/api/cashReconciliation.ts
import { api } from './client'

export interface PettyCashEntry {
  id: string
  amount: string
  description: string
  created_at: string
}

export interface CashReconciliation {
  id: string
  business_date: string      // YYYY-MM-DD
  opening_balance: string
  cash_in: string
  petty_cash_net: string
  expected_balance: string
  counted_balance: string | null
  deposit_amount: string
  closing_balance: string | null
  variance: string | null
  variance_note: string | null
  status: 'open' | 'closed'
  closed_at: string | null
  petty_cash_entries: PettyCashEntry[]
}

export function getCurrentReconciliation(): Promise<CashReconciliation> {
  return api.get<CashReconciliation>('/cash-reconciliation/current')
}

export function listReconciliations(): Promise<CashReconciliation[]> {
  return api.get<CashReconciliation[]>('/cash-reconciliation')
}

export function getReconciliation(date: string): Promise<CashReconciliation> {
  return api.get<CashReconciliation>(`/cash-reconciliation/${date}`)
}

export function openPeriod(businessDate: string, openingBalance?: string): Promise<CashReconciliation> {
  return api.post<CashReconciliation>('/cash-reconciliation', {
    business_date: businessDate,
    ...(openingBalance !== undefined ? { opening_balance: openingBalance } : {}),
  })
}

export function closePeriod(
  date: string,
  body: { counted_balance: string; deposit_amount: string; variance_note: string },
): Promise<CashReconciliation> {
  return api.post<CashReconciliation>(`/cash-reconciliation/${date}/close`, body)
}

export function addPettyCash(
  date: string,
  body: { amount: string; description: string },
): Promise<CashReconciliation> {
  return api.post<CashReconciliation>(`/cash-reconciliation/${date}/petty-cash`, body)
}

export function deletePettyCash(date: string, entryId: string): Promise<CashReconciliation> {
  return api.delete<CashReconciliation>(`/cash-reconciliation/${date}/petty-cash/${entryId}`)
}
